feat: zookeeper logs for humans

3Hren opened the pull request:
This PR redirects all zookeeper client logs to our logging system. No more stderr Hell!

@antmat TAL.
@noxiouz @toshik @borislitv @agodin you'd may also be interested.

3Hren committed:
feat: zookeeper logs for humans

antmat commented on unicorn/src/repository/zookeeper.cpp:
```diff
@@ -0,0 +1,124 @@
+    cocaine::logging::logger_t* log = static_cast<cocaine::logging::logger_t*>(handler);
+
+    std::string message(buf, size);
+    while (message.size() > 0 && message.back() == '\n') {
```
boost::trim?

3Hren replied:
Ok.

antmat commented on unicorn/src/repository/zookeeper.cpp:
```diff
@@ -0,0 +1,124 @@
+        message.resize(message.size() - 1);
+    }
+
+    auto pos = message.find("ZOO_");
```
I'm not sure, but it seems that 
`
static std::string prefix("ZOO_");
auto pos = message.find(prefix);
`
would be more efficient

3Hren replied:
There is `const char*` overload.

antmat replied:
yep, but in case of static string you always know the size. Whatever.

antmat commented on unicorn/src/repository/zookeeper.cpp:
```diff
@@ -0,0 +1,124 @@
+        return 0;
+    }
+
+    int severity = 0;
```
I would add any unknown prefixes to higher severity than DEBUG

3Hren replied:
But there are no more levels: https://github.com/apache/zookeeper/blob/master/src/c/include/zookeeper.h#L138.

antmat replied:
so std::terminate in default is safe?

antmat commented on unicorn/src/repository/zookeeper.cpp:
```diff
@@ -0,0 +1,124 @@
+        ;
+    }
+
+    pos = message.find(": ");
```
It's better to look for this colon after prefix just in case(and it will be more efficient).

antmat commented on unicorn/src/repository/zookeeper.cpp:
```diff
@@ -0,0 +1,124 @@
+
+    pos = message.find(": ");
+    message = message.substr(pos + 2);
+    if (message.size() > 0) {
```
empty() is more straitforward

3Hren replied:
Then it will be inverted condition. Inverted conditions are bad.

antmat replied:
Not really. inverted condition in methods resulting in double inversion are bad (e.g. !not_found())

antmat commented on unicorn/src/repository/zookeeper.cpp:
```diff
@@ -0,0 +1,124 @@
+static int do_log(void* handler, const char* buf, size_t size) {
+    cocaine::logging::logger_t* log = static_cast<cocaine::logging::logger_t*>(handler);
+
+    std::string message(buf, size);
```
why not string view? I thought you have it in blackhole

3Hren replied:
String view is immutable. We need mutations.

antmat replied:
Your only real mutation(except cutting) is tolower :-))) 
Anyway we don't expect much logs from ZK.

antmat commented on unicorn/src/repository/zookeeper.cpp:
```diff
@@ -0,0 +1,124 @@
+    return size;
+}
+
+#ifdef __linux__
```
you can make one typedef for size_type and use it

antmat commented on unicorn/src/repository/zookeeper.cpp:
```diff
@@ -0,0 +1,124 @@
+zookeeper_factory_t::ptr_type
+zookeeper_factory_t::get(context_t& context, const std::string& name, const dynamic_t& args) {
+    if (initialized.test_and_set() == false) {
+        fh = init_logging(context.log(cocaine::format("zookeeper")).release());
```
unused variable

antmat commented on unicorn/src/repository/zookeeper.cpp:
```diff
@@ -0,0 +1,124 @@
+
+zookeeper_factory_t::zookeeper_factory_t() {}
+zookeeper_factory_t::~zookeeper_factory_t() {
+    ::fclose(fh);
```
I think we need to redirect zookeeper back to stderr here, just in case.

3Hren replied:
Good catch.

antmat reviewed (approved):
Great job! I didn't know about fopencookie function, looks cool, if it works.
Some notice:
  * you do initialize global state of zookeeper library from nonstatic factory. It looks like it have nothing to do with the factory itself, but as far as we have only one factory in our repository it's more or less ok.

Toshik commented:
Guys, I do not even know what is this all about...

> ...@toshik ... you'd may also be interested.

3Hren commented:
@Toshik whatever. Review!

Toshik commented:
@3Hren Чего ревью-то? Попутал меня, видать, с кем-то ))

3Hren commented:
@Toshik да я уже понял, что попутал. Все равно ревьюй! :)

antmat commented on unicorn/src/repository/zookeeper.cpp:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    auto pos = message.find("ZOO_");
+    if (pos == std::string::npos || pos + 4 >= message.size()) {
```
BTW, why dont' you log such strange messages, but silently eat them?

3Hren committed:
refactor: trimming the right wa

3Hren committed:
refactor: code style

3Hren committed:
fix: proper destruction

3Hren committed:
refactor: global once initialization

antmat commented on unicorn/src/repository/zookeeper.cpp:
```diff
@@ -0,0 +1,142 @@
+
+zookeeper_factory_t::ptr_type
+zookeeper_factory_t::get(context_t& context, const std::string& name, const dynamic_t& args) {
+    std::call_once(init_once, [&] {
```
you introduced a segfault here. Outer logger will go out of scope, while initializer_t will be in scope and zk can still use logger pointing to released root_logger

3Hren replied:
Better ideas?

3Hren committed:
refactor: move here, move there

3Hren committed:
refactor: self explained termination

3Hren committed:
fix: well, fix almost zero-probability SEGFAULT

Since the ZK factory is private type, it’s ok for us to maintain once
initialization manually without static initialization/destruction magic.

3Hren committed:
fix: reset ZK logging system before closing
